class.c

```diff
@@ -0,0 +1,1033 @@
+/*    class.c
+ *
+ *    Copyright (C) 2022 by Paul Evans and others
+ *
+ *    You may distribute under the terms of either the GNU General Public
+ *    License or the Artistic License, as specified in the README file.
+ *
+ */
+
+/* This file contains the code that implements perl's new `use feature 'class'`
+ * object model
+ */
+
+#include "EXTERN.h"
+#define PERL_IN_CLASS_C
+#include "perl.h"
+
+#include "XSUB.h"
+
+enum {
+    PADIX_SELF   = 1,
+    PADIX_PARAMS = 2,
+};
+
+void
+Perl_croak_kw_unless_class(pTHX_ const char *kw)
+{
+    PERL_ARGS_ASSERT_CROAK_KW_UNLESS_CLASS;
+
+    if(!HvSTASH_IS_CLASS(PL_curstash))
+        croak("Cannot '%s' outside of a 'class'", kw);
+}
+
+#define newSVobject(fieldcount)  Perl_newSVobject(aTHX_ fieldcount)
+SV *
+Perl_newSVobject(pTHX_ Size_t fieldcount)
+{
+    SV *sv = newSV_type(SVt_PVOBJ);
+
+    Newx(ObjectFIELDS(sv), fieldcount, SV *);
+    ObjectMAXFIELD(sv) = fieldcount - 1;
+
+    Zero(ObjectFIELDS(sv), fieldcount, SV *);
+
+    return sv;
+}
+
+PP(pp_initfield)
+{
+    dSP;
+    UNOP_AUX_item *aux = cUNOP_AUX->op_aux;
+
+    SV *self = PAD_SVl(PADIX_SELF);
+    assert(SvTYPE(SvRV(self)) == SVt_PVOBJ);
+    SV *instance = SvRV(self);
+
+    SV **fields = ObjectFIELDS(instance);
+
+    PADOFFSET fieldix = aux[0].uv;
+
+    SV *val = NULL;
+
+    switch(PL_op->op_private & (OPpINITFIELD_AV|OPpINITFIELD_HV)) {
+        case 0:
+            if(PL_op->op_flags & OPf_STACKED)
+                val = newSVsv(POPs);
+            else
+                val = newSV(0);
+            break;
+
+        case OPpINITFIELD_AV:
+        {
+            AV *av = newAV();
+            if(PL_op->op_flags & OPf_STACKED) {
+                SV **svp = PL_stack_base + POPMARK + 1;
+                STRLEN count = SP - svp + 1;
+
+                av_extend(av, count);
```

Comment:
More efficient to directly create the AV with SV** array of size count. Maybe like this:
```
    AV *av;
    if(PL_op->op_flags & OPf_STACKED) {
￼        SV **svp = PL_stack_base + POPMARK + 1;
￼        STRLEN count = SP - svp;
￼
￼        av = newAV_alloc_x(count);
￼        while(svp <= SP) {
￼            av_push_simple(av, newSVsv(*svp));
￼            svp++;
￼        }
    } else
        av = newAV();
```
(I haven't had coffee yet this morning, `count` size may need adjusting.)

Reply:
Ahyes; that looks a bit better. Count was fine - it's count, not max index. So no off-by-one adjustment needed.
Have force-pushed.